src/auth/googleauth.ts

```diff
@@ -745,7 +745,16 @@ export class GoogleAuth {
    */
   async getRequestHeaders(url?: string) {
     const client = await this.getClient();
-    return client.getRequestHeaders(url);
+    let headers = client.getRequestHeaders(url);
+    // quota_project, stored in application_default_credentials.json, is set in
+    // the x-goog-user-project header, to indicate an alternate account for
+    // billing and quota:
+    if (this.jsonContent && this.jsonContent.quota_project) {
+      headers = Object.assign({}, headers, {
+        'x-goog-user-project': this.jsonContent.quota_project,
+      });
```

Comment:
A comment about why it's not the other way around, i.e.
`Object.assign({}, ..., headers)`

Reply:
If `getRequestHeaders` always returns an object, the whole `Object.assign` can be removed and replaced with a simple assignment :)

Reply:
I tend to agree that `headers` should take precedence if set explicitly, I've updated the PR and added a comment.

Reply:
I've noticed a few bugs in our auth library in my time working on it, caused by passing an object by reference and than modifying the object, e.g., deleting a key.

I've been being defensive by creating shallow copies with `Object.assign({})`, something worth thinking about in API design would be starting to `Object.freeze` objects that we don't wish for users to modify, at which point it forces people to make copies, but you can pass the object around if you do not need to mutate it.